app/src/main/java/com/keylesspalace/tusky/OrderableListPreferenceActivity.kt

```diff
@@ -0,0 +1,448 @@
+        editText.requestFocus()
+    }
+
+    private fun showSelectListDialog() {
+        val adapter = ListSelectionAdapter(this)
+
+        val statusLayout = LinearLayout(this)
+        statusLayout.gravity = Gravity.CENTER
+        val progress = ProgressBar(this)
+        val preferredPadding = getDimension(this, androidx.appcompat.R.attr.dialogPreferredPadding)
+        progress.setPadding(preferredPadding, 0, preferredPadding, 0)
+        progress.visible(false)
+
+        val noListsText = TextView(this)
+        noListsText.setPadding(preferredPadding, 0, preferredPadding, 0)
+        noListsText.text = getText(R.string.select_list_empty)
+        noListsText.visible(false)
+
+        statusLayout.addView(progress)
+        statusLayout.addView(noListsText)
+
+        val dialogBuilder = AlertDialog.Builder(this)
+            .setTitle(R.string.select_list_title)
+            .setNeutralButton(R.string.select_list_manage) { _, _ ->
+                val listIntent = Intent(applicationContext, ListsActivity::class.java)
+                startActivity(listIntent)
+            }
+            .setNegativeButton(android.R.string.cancel, null)
+            .setView(statusLayout)
+            .setAdapter(adapter) { _, position ->
+                val list = adapter.getItem(position)
+                val newTab = createTabDataFromId(LIST, listOf(list!!.id, list.title))
+                currentTabs.add(newTab)
+                currentTabsAdapter.notifyItemInserted(currentTabs.size - 1)
+                updateAvailableTabs()
+                saveList(currentTabs)
+            }
+
+        val showProgressBarJob = getProgressBarJob(progress, 500)
+        showProgressBarJob.start()
+
+        val dialog = dialogBuilder.show()
+
+        lifecycleScope.launch {
+            mastodonApi.getLists().fold(
+                { lists ->
+                    showProgressBarJob.cancel()
+                    adapter.addAll(lists)
+                    if (lists.isEmpty()) {
+                        noListsText.show()
+                    }
+                },
+                { throwable ->
+                    dialog.hide()
+                    Log.e("TabPreferenceActivity", "failed to load lists", throwable)
+                    Snackbar.make(binding.root, R.string.error_list_load, Snackbar.LENGTH_LONG).show()
+                }
+            )
+        }
+    }
+
+    private fun getProgressBarJob(progressView: View, delayMs: Long) = this.lifecycleScope.launch(
+        start = CoroutineStart.LAZY
+    ) {
+        try {
+            delay(delayMs)
+            progressView.show()
+            awaitCancellation()
+        } finally {
+            progressView.hide()
+        }
+    }
+
+    private fun validateHashtag(input: CharSequence?): Boolean {
+        val trimmedInput = input?.trim() ?: ""
+        return trimmedInput.isNotEmpty() && hashtagRegex.matcher(trimmedInput).matches()
+    }
+
+    private fun updateAvailableTabs() {
```

Comment:
This method could be significantly simplified if there was a parameter to check if multiple entries in a menu are allowed on `TabData`